src/linear_algebra/dual.lean

```diff
@@ -854,25 +981,251 @@ begin
+  (f : M →ₗ[R] M') (hf : function.surjective f.range.subtype.dual_map) :
+  f.dual_map.range = f.ker.dual_annihilator :=
+begin
+  have rr_surj : function.surjective f.range_restrict,
+  { rw [← linear_map.range_eq_top, linear_map.range_range_restrict] },
+  have := range_dual_map_eq_dual_annihilator_ker_of_surjective f.range_restrict rr_surj,
+  convert this using 1,
+  { change ((submodule.subtype f.range).comp f.range_restrict).dual_map.range = _,
+    rw [← linear_map.dual_map_comp_dual_map, linear_map.range_comp_of_range_eq_top],
+    rwa linear_map.range_eq_top, },
+  { apply congr_arg,
+    exact (linear_map.ker_range_restrict f).symm, },
+end
+
+end linear_map
+
+end comm_ring
+
+section vector_space
 
 variables {K : Type*} [field K] {V₁ : Type*} {V₂ : Type*}
 variables [add_comm_group V₁] [module K V₁] [add_comm_group V₂] [module K V₂]
 
-open finite_dimensional
+namespace linear_map
+
+lemma dual_pairing_nondegenerate : (dual_pairing K V₁).nondegenerate :=
+⟨separating_left_iff_ker_eq_bot.mpr ker_id, λ x, (forall_dual_apply_eq_zero_iff K x).mp⟩
+
+lemma dual_map_surjective_of_injective {f : V₁ →ₗ[K] V₂} (hf : function.injective f) :
+  function.surjective f.dual_map :=
+begin
+  intro φ,
+  let f' := linear_equiv.of_injective f hf,
+  use subspace.dual_lift (range f) (f'.symm.dual_map φ),
+  ext x,
+  rw [linear_map.dual_map_apply, subspace.dual_lift_of_mem (mem_range_self f x),
+    linear_equiv.dual_map_apply],
+  congr' 1,
+  exact linear_equiv.symm_apply_apply f' x,
+end
+
+lemma range_dual_map_eq_dual_annihilator_ker (f : V₁ →ₗ[K] V₂) :
+  f.dual_map.range = f.ker.dual_annihilator :=
+range_dual_map_eq_dual_annihilator_ker_of_subtype_range_surjective f $
+  dual_map_surjective_of_injective (range f).injective_subtype
+
+end linear_map
+
+namespace subspace
+open submodule
+
+lemma dual_pairing_eq (W : subspace K V₁) :
+  W.dual_pairing = W.quot_annihilator_equiv.to_linear_map :=
+by { ext, refl }
+
+lemma dual_pairing_nondegenerate (W : subspace K V₁) : W.dual_pairing.nondegenerate :=
+begin
+  split,
+  { rw [linear_map.separating_left_iff_ker_eq_bot, dual_pairing_eq],
+    apply linear_equiv.ker, },
+  { intros x h,
+    rw ← forall_dual_apply_eq_zero_iff K x,
+    intro φ,
+    simpa only [submodule.dual_pairing_apply, dual_lift_of_subtype]
+      using h (submodule.quotient.mk (W.dual_lift φ)), }
+end
+
+lemma dual_copairing_nondegenerate (W : submodule K V₁) : W.dual_copairing.nondegenerate :=
+begin
+  split,
+  { rw [linear_map.separating_left_iff_ker_eq_bot, dual_copairing_eq],
+    apply linear_equiv.ker, },
+  { rintro ⟨x⟩,
+    simp only [quotient.quot_mk_eq_mk, dual_copairing_apply, quotient.mk_eq_zero],
+    rw [← forall_mem_dual_annihilator_apply_eq_zero_iff, set_like.forall],
+    exact id, }
+end
+
+-- Argument from https://math.stackexchange.com/a/2423263/172988
+lemma dual_annihilator_inf_eq (W W' : subspace K V₁) :
+  (W ⊓ W').dual_annihilator = W.dual_annihilator ⊔ W'.dual_annihilator :=
```

Comment:
It took me some time to come up with a (probably the simplest) counterexample over ℤ: if `W` and `W'` are submodules of `ℤ ⊕ ℤ` spanned by `(2,1)` and `(1,2)` respectively, there annihilators are spanned by `(1,-2)` and `(2,-1)` respectively. However, `ann (W ⊓ W') = ann 0 = ⊤`, but `ann W ⊔ ann W'` has index 3 in `ℤ ⊕ ℤ`, since the determinant of [[1,-2],[2,-1]] is 3.